ICE - from Clang/ItaniumMangle

I compiled the latest develop branch with clang-11 (llvmorg-11.0.0-rc2), as well as with 10.0, both with CUDA 10.1, but in both cases, I am getting an ICE out from the ItaniumMangle module as below. I don't even know why Itanium is coming up, I only set llvm targets to X86 and NVPTX... I can compile other C++ and CUDA code okay. Any ideas how to work around this?

> syclcc-clang -O3 --hipsycl-gpu-arch=sm_60 test.cpp
> Can't mangle a deduction guide name!
> UNREACHABLE executed at /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:1509!
> PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.	Program arguments: /home/ireguly/projects/llvm-install/bin/clang-11 -cc1 -triple nvptx64-nvidia-cuda -aux-triple x86_64-unknown-linux-gnu -S -disable-free -main-file-name test.cpp -mrelocation-model static -mframe-pointer=all -fno-rounding-math -fno-verbose-asm -no-integrated-as -aux-target-cpu x86-64 -fcuda-is-device -mlink-builtin-bitcode /usr/local/cuda/nvvm/libdevice/libdevice.10.bc -target-feature +ptx64 -target-sdk-version=10.1 -target-cpu sm_60 -fno-split-dwarf-inlining -debugger-tuning=gdb -resource-dir /home/ireguly/projects/llvm-install/lib/clang/11.0.0 -internal-isystem /home/ireguly/projects/llvm-install/lib/clang/11.0.0/include/cuda_wrappers -internal-isystem /usr/local/cuda/include -include __clang_cuda_runtime_wrapper.h -D HIPSYCL_CLANG -I /home/ireguly/projects/hipSYCL/install4/bin/../include/ -I /home/ireguly/projects/hipSYCL/install4/bin/../include/hipSYCL/contrib -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8 -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/x86_64-redhat-linux -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/backward -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8 -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/x86_64-redhat-linux -internal-isystem /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/backward -internal-isystem /usr/local/include -internal-isystem /home/ireguly/projects/llvm-install/lib/clang/11.0.0/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -internal-isystem /usr/local/include -internal-isystem /home/ireguly/projects/llvm-install/lib/clang/11.0.0/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -O3 -std=c++17 -fdeprecated-macro -fno-dwarf-directory-asm -fno-autolink -fdebug-compilation-dir /home/ireguly -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -vectorize-loops -vectorize-slp -load /home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so -o /tmp/test-c04f4e.s -x cuda test.cpp
> 1.	/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/cstdlib:121:1: current parser token 'extern'
> 2.	/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_pair.h:218:26: Mangling declaration 'std::<deduction guide for pair>'
>  #0 0x00007f4cbdf3b718 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/ireguly/projects/llvm-project/llvm/lib/Support/Unix/Signals.inc:564:22
>  #1 0x00007f4cbdf3b7ab PrintStackTraceSignalHandler(void*) /home/ireguly/projects/llvm-project/llvm/lib/Support/Unix/Signals.inc:625:1
>  #2 0x00007f4cbdf3983d llvm::sys::RunSignalHandlers() /home/ireguly/projects/llvm-project/llvm/lib/Support/Signals.cpp:68:20
>  #3 0x00007f4cbdf3b15b SignalHandler(int) /home/ireguly/projects/llvm-project/llvm/lib/Support/Unix/Signals.inc:406:1
>  #4 0x00007f4ccd080d80 __restore_rt (/lib64/libpthread.so.0+0x12d80)
>  #5 0x00007f4cbac1793f raise (/lib64/libc.so.6+0x3793f)
>  #6 0x00007f4cbac01c95 abort (/lib64/libc.so.6+0x21c95)
>  #7 0x00007f4cbddeb412 bindingsErrorHandler(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /home/ireguly/projects/llvm-project/llvm/lib/Support/ErrorHandling.cpp:219:55
>  #8 0x00007f4cc607134c (anonymous namespace)::CXXNameMangler::mangleUnqualifiedName(clang::GlobalDecl, clang::DeclarationName, unsigned int, llvm::SmallVector<llvm::StringRef, 4u> const*) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:1512:5
>  #9 0x00007f4cc606e3fc (anonymous namespace)::CXXNameMangler::mangleUnqualifiedName(clang::GlobalDecl, llvm::SmallVector<llvm::StringRef, 4u> const*) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:475:3
> #10 0x00007f4cc606faea (anonymous namespace)::CXXNameMangler::mangleUnscopedName(clang::GlobalDecl, llvm::SmallVector<llvm::StringRef, 4u> const*) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:970:1
> #11 0x00007f4cc606f6aa (anonymous namespace)::CXXNameMangler::mangleNameWithAbiTags(clang::GlobalDecl, llvm::SmallVector<llvm::StringRef, 4u> const*) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:906:5
> #12 0x00007f4cc606f466 (anonymous namespace)::CXXNameMangler::mangleName(clang::GlobalDecl) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:857:1
> #13 0x00007f4cc606ea25 (anonymous namespace)::CXXNameMangler::mangleFunctionEncoding(clang::GlobalDecl) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:678:35
> #14 0x00007f4cc606e793 (anonymous namespace)::CXXNameMangler::mangle(clang::GlobalDecl) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:662:1
> #15 0x00007f4cc607f48b (anonymous namespace)::ItaniumMangleContextImpl::mangleCXXName(clang::GlobalDecl, llvm::raw_ostream&) /home/ireguly/projects/llvm-project/clang/lib/AST/ItaniumMangle.cpp:5143:39
> #16 0x00007f4cc60b3df1 clang::MangleContext::mangleName(clang::GlobalDecl, llvm::raw_ostream&) /home/ireguly/projects/llvm-project/clang/lib/AST/Mangle.cpp:181:5
> #17 0x00007f4cba141967 hipsycl::compiler::FrontendASTVisitor::getMangledName[abi:cxx11](clang::FunctionDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x5b967)
> #18 0x00007f4cba140df7 hipsycl::compiler::FrontendASTVisitor::processFunctionDecl(clang::FunctionDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x5adf7)
> #19 0x00007f4cba12c326 clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseCXXDeductionGuideDecl(clang::CXXDeductionGuideDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x46326)
> #20 0x00007f4cba127ecd clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseDecl(clang::Decl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x41ecd)
> #21 0x00007f4cba12a1c1 clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseFunctionTemplateDecl(clang::FunctionTemplateDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x441c1)
> #22 0x00007f4cba127c8d clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseDecl(clang::Decl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x41c8d)
> #23 0x00007f4cba1291eb clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseNamespaceDecl(clang::NamespaceDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x431eb)
> #24 0x00007f4cba127b55 clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseDecl(clang::Decl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x41b55)
> #25 0x00007f4cba128fdb clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseLinkageSpecDecl(clang::LinkageSpecDecl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x42fdb)
> #26 0x00007f4cba127b25 clang::RecursiveASTVisitor<hipsycl::compiler::FrontendASTVisitor>::TraverseDecl(clang::Decl*) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x41b25)
> #27 0x00007f4cba1271eb hipsycl::compiler::FrontendASTConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/home/ireguly/projects/hipSYCL/install4/bin/../lib/libhipSYCL_clang.so+0x411eb)
> #28 0x00007f4cc8a9ed53 clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/ireguly/projects/llvm-project/clang/lib/Frontend/MultiplexConsumer.cpp:271:25
> #29 0x00007f4cc5a2dcc0 clang::ParseAST(clang::Sema&, bool, bool) /home/ireguly/projects/llvm-project/clang/lib/Parse/ParseAST.cpp:162:20
> #30 0x00007f4cc8a6c785 clang::ASTFrontendAction::ExecuteAction() /home/ireguly/projects/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1057:11
> #31 0x00007f4cc77bcb4f clang::CodeGenAction::ExecuteAction() /home/ireguly/projects/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:1185:1
> #32 0x00007f4cc8a6c0ed clang::FrontendAction::Execute() /home/ireguly/projects/llvm-project/clang/lib/Frontend/FrontendAction.cpp:954:38
> #33 0x00007f4cc8a13690 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ireguly/projects/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:984:42
> #34 0x00007f4cc8b3bed5 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ireguly/projects/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:278:38
> #35 0x000000000044ac74 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ireguly/projects/llvm-project/clang/tools/driver/cc1_main.cpp:240:40
> #36 0x0000000000440cab ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /home/ireguly/projects/llvm-project/clang/tools/driver/driver.cpp:330:20
> #37 0x00000000004413ac main /home/ireguly/projects/llvm-project/clang/tools/driver/driver.cpp:407:26
> #38 0x00007f4cbac03813 __libc_start_main (/lib64/libc.so.6+0x23813)
> #39 0x000000000043f53e _start (/home/ireguly/projects/llvm-install/bin/clang-11+0x43f53e)
> clang-11: error: unable to execute command: Aborted (core dumped)
> clang-11: error: clang frontend command failed due to signal (use -v to see invocation)
> clang version 11.0.0 (https://github.com/llvm/llvm-project.git 414f32a9e862b11f51063b75729278f8d81b12e9)
> Target: x86_64-unknown-linux-gnu
> Thread model: posix
> InstalledDir: /home/ireguly/projects/llvm-install/bin
> clang-11: note: diagnostic msg:
> ********************
> 
> PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
> Preprocessed source(s) and associated run script(s) are located at:
> clang-11: note: diagnostic msg: /tmp/test-4325e9.cu
> clang-11: note: diagnostic msg: /tmp/test-ba90ad.cu
> clang-11: note: diagnostic msg: /tmp/test-4325e9.sh
> clang-11: note: diagnostic msg:
> 
> ********************

test.cpp is:

> #include <CL/sycl.hpp>
> #include <iostream>
> 
> int main() {
>   cl::sycl::gpu_selector Selector;
> 
>   // Creating buffer of 4 ints to be used inside the kernel code
>   cl::sycl::buffer<cl::sycl::cl_int, 1> Buffer(4);
> 
>   // Creating SYCL queue
>   cl::sycl::queue Queue(Selector);
>   std::cout << "Running on " << Queue.get_device().get_info<cl::sycl::info::device::name>() << "\n";
>   // Size of index space for kernel
>   cl::sycl::range<1> NumOfWorkItems{Buffer.get_count()};
> 
>   // Submitting command group(work) to queue
>   Queue.submit([&](cl::sycl::handler &cgh) {
>     // Getting write only access to the buffer on a device
>     auto Accessor = Buffer.get_access<cl::sycl::access::mode::write>(cgh);
>     // Executing kernel
>     cgh.parallel_for < class FillBuffer >(
>         NumOfWorkItems, [=](cl::sycl::id<1> WIid) {
>           // Fill buffer with indexes
>           Accessor[WIid] = (cl::sycl::cl_int)WIid.get(0);
>         });
>   });
> 
>   // Getting read only access to the buffer on the host.
>   // Implicit barrier waiting for queue to complete the work.
>   const auto HostAccessor = Buffer.get_access<cl::sycl::access::mode::read>();
> 
>   // Check the results
>   bool MismatchFound = false;
>   for (size_t I = 0; I < Buffer.get_count(); ++I) {
>     if (HostAccessor[I] != I) {
>       std::cout << "The result is incorrect for element: " << I
>                 << " , expected: " << I << " , got: " << HostAccessor[I]
>                 << std::endl;
>       MismatchFound = true;
>     }
>   }
> 
>   if (!MismatchFound) {
>     std::cout << "The results are correct!" << std::endl;
>   }
> 
>   return MismatchFound;
> }
>

I am also seeing this issue when building for ROCm with either 'gfx803' or 'gfx900' in `develop`/ 5216d091cf92a

I'm on Debian Buster (4.19.0-5-amd64 SMP Debian 4.19.37-3) and have tested with the buster-backports version of Clang 8, and source builds of Clang 9.0.1, 10, and the LLVM GitHub. My host compiler is GCC 8.3.0. ROCm installation is 2.9.6

Can confirm that after merging in #303 into 5216d091cf92a, the ROCm compilation appears to be working again on Clang >=9 (less some minor ROCm environment issues on my machine)
[LLVM GitHub @ 51d1d58](https://github.com/llvm/llvm-project/commit/51d1d585e5838ea0f02f1271f7543c4e43639969): Works
Clang 10: Works
Clang 9.0.1: Works
Clang 8: lots of unrelated `_Float16` errors:
```
/opt/rocm/include/hip/hcc_detail/hip_fp16_math_fwd.h:33:64: error:
      _Float16 is not supported on this target
  ...__attribute__((const)) _Float16 __ocml_ceil_f16(_Float16);
```

Sidenote: [The ROCm install notes](https://github.com/illuhad/hipSYCL/blob/develop/doc/install-rocm.md) indicate that as far back as 2.6 is supported. We had to do some workaround on our 2.9.6 install to make things work.:
1. apply https://github.com/ROCm-Developer-Tools/HIP/commit/08d9759ebade199512a7d8dec319cfc098a556b1 to resolve a header ordering issue with the implementation of STL `algorithm`
2. Change [hipSYCL's embedded HIP library link flag](https://github.com/illuhad/hipSYCL/blob/develop/CMakeLists.txt#L134) to point to `libhip_hcc.so` instead of `libamdhip64.so`, since the latter didn't exist. It's not exactly what the modern ROCm does ([create a symlink the other direction](https://github.com/ROCm-Developer-Tools/HIP/blob/master/vdi/CMakeLists.txt#L170)), but sufficient.

--- a/include/hipSYCL/compiler/Frontend.hpp
+++ b/include/hipSYCL/compiler/Frontend.hpp
@@ -210,17 +210,14 @@ inline std::string buildKernelName(clang::TemplateArgument SyclTagType) {
 class FrontendASTVisitor : public clang::RecursiveASTVisitor<FrontendASTVisitor>
 {
   clang::CompilerInstance &Instance;
-  clang::MangleContext *MangleContext;
+  
 public:
   FrontendASTVisitor(clang::CompilerInstance &instance)
-      : Instance{instance},
-        MangleContext{Instance.getASTContext().createMangleContext()}
+      : Instance{instance}
   {}
 
   ~FrontendASTVisitor()
-  {
-    delete this->MangleContext;
-  }
+  {}
   
   bool shouldVisitTemplateInstantiations() const { return true; }
 
@@ -382,7 +379,6 @@ public:
       {
         HIPSYCL_DEBUG_INFO << "AST processing: Marking function as __host__ __device__: "
                            << RD->getQualifiedNameAsString() << std::endl;
-        CompilationStateManager::getASTPassState().addImplicitHostDeviceFunction(getMangledName(RD));
         markAsHostDevice(RD);
         if (!RD->hasAttr<clang::CUDAHostAttr>())
           RD->addAttr(clang::CUDAHostAttr::CreateImplicit(Instance.getASTContext()));
@@ -455,20 +451,9 @@ private:
       }
     }
   
-    std::string MangledName = getMangledName(f);
-    if(CustomAttributes::SyclKernel.isAttachedTo(f))
-    {
+    
+    if(CustomAttributes::SyclKernel.isAttachedTo(f)){
       markAsKernel(f); 
-      CompilationStateManager::getASTPassState().addKernelFunction(MangledName);
-    }
-    else if(f->hasAttr<clang::CUDADeviceAttr>())
-    {
-      if(!f->getAttr<clang::CUDADeviceAttr>()->isImplicit())
-        CompilationStateManager::getASTPassState().addExplicitDeviceFunction(MangledName);
-    }
-    else if(f->hasAttr<clang::CUDAGlobalAttr>())
-    {
-      CompilationStateManager::getASTPassState().addKernelFunction(MangledName);
     }
   }
 
@@ -554,21 +539,6 @@ private:
     }
   }
 
-  std::string getMangledName(clang::FunctionDecl* decl)
-  {
-    if (!MangleContext->shouldMangleDeclName(decl)) {
-      return decl->getNameInfo().getName().getAsString();
-    }
-
-    std::string mangledName;
-    llvm::raw_string_ostream ostream(mangledName);
-
-    MangleContext->mangleName(decl, ostream);
-
-    ostream.flush();
-
-    return mangledName;
-  }
 };
 
 

